File: Raspberry Pi/PWM Examples/python/pwm_example.py
#!/usr/bin/python3
#**
#******************************************************************************
#* @file    /pwm_examples/python/pwm_example.py
#* @author  Adrian Wojcik
#* @version V1.0
#* @date    15-Mar-2020
#* @brief   Raspberry Pi PWM control: Python 3 with RPi.GPIO lib
#******************************************************************************

import time
import sys
import select
import numpy

try:
	import RPi.GPIO as GPIO
except RuntimeError:
	print("Error importing RPi.GPIO! Use 'sudo' to run your script")
  
timeout = 1	
i = ''	
duty = 0 # [%]
freq = 500 # [Hz]
	
# Pin Definitons:
pwmPin = 12    #< LED: Physical pin 12, BCM GPIO18	
	
GPIO.setmode(GPIO.BOARD)
GPIO.setup(pwmPin, GPIO.OUT)

p = GPIO.PWM(pwmPin, freq)  
p.start(duty)

print("Press ENTER to exit.")

while not i:
	# Waiting for I/O completion
	i, o, e = select.select( [sys.stdin], [], [], timeout )

	if (i):
		sys.stdin.readline();
		p.stop()
		GPIO.cleanup() # cleanup all GPIO
		exit()
		
	for d in numpy.arange(1, 100, 1):
		#print(d)
		p.ChangeDutyCycle(d)
		time.sleep(0.02)
	for d in numpy.arange(100, -1, -1):
		#print(d)
		p.ChangeDutyCycle(d)
		time.sleep(0.02)

File: Raspberry Pi/PWM Examples/c++/pwm_example.cpp
/**
 ******************************************************************************
 * @file    /pwm_examples/cpp/pwm_example.cpp
 * @author  Adrian Wojcik
 * @version V1.0
 * @date    15-Mar-2020
 * @brief   Raspberry Pi PWM control: C++ with wiringPi lib
 ******************************************************************************
 */

#include <iostream>
#include <future>
#include <thread>
#include <chrono>
#include <wiringPi.h>

int main(int argc, char *argv[]) 
{
	const int pwmPin = 1; //< Red LED: Physical pin 12, BCM GPIO18, and WiringPi pin 1.
	const int range = 200;
	const int clock = 192;
	int step = 1;

	std::chrono::milliseconds timeout(1000);
	std::chrono::milliseconds delay(10);
	std::future<int> async_getchar = std::async(std::getchar);

	wiringPiSetup();

	pinMode(pwmPin, PWM_OUTPUT);
	
	pwmSetMode(PWM_MODE_MS);
	pwmSetRange(range);
	pwmSetClock(clock);
	
	pwmWrite(pwmPin, 0) ;
	
	std::cout << "Press ENTER to exit." << std::endl;
	
	while(1) 
	{	
		if(async_getchar.wait_for(timeout) == std::future_status::ready)
		{
			async_getchar.get();		
			break;
		}	

		for(int d = 0 ; d <= range ; d+=step)
		{
			pwmWrite(pwmPin, d) ;
			std::this_thread::sleep_for(delay);	
		}

		for(int d = range ; d >= 0 ; d-=step)
		{
			pwmWrite(pwmPin, d) ;
			std::this_thread::sleep_for(delay);	;
		}
	}
    
	return 0;
}